pkg/schema/definitions/handler.go

```diff
@@ -0,0 +1,197 @@
+	"time"
+
+	"github.com/rancher/apiserver/pkg/apierror"
+	"github.com/rancher/apiserver/pkg/types"
+	"github.com/rancher/steve/pkg/schema/converter"
+	"github.com/rancher/wrangler/pkg/schemas/validation"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+var (
+	internalServerErrorCode = validation.ErrorCode{
+		Status: http.StatusInternalServerError,
+		Code:   "InternalServerError",
+	}
+	notRefreshedErrorCode = validation.ErrorCode{
+		Status: http.StatusServiceUnavailable,
+		Code:   "SchemasNotRefreshed",
+	}
+)
+
+// schemaDefinitionHandler is a byID handler for a specific schema, which provides field definitions for all schemas.
+// Does not implement any method allowing a caller to list definitions for all schemas.
+type schemaDefinitionHandler struct {
+	sync.RWMutex
+
+	// lastRefresh is the last time that the handler retrieved models from kubernetes.
+	lastRefresh time.Time
+	// refreshStale is the duration between lastRefresh and the next refresh of models.
+	refreshStale time.Duration
+	// client is the discovery client used to get the groups/resources/fields from kubernetes.
+	client discovery.DiscoveryInterface
+	// models are the cached models from the last response from kubernetes.
+	models *proto.Models
+	// schemaToModel is a map of the schema name to the model for that schema. Can be used to load the
+	// top-level definition for a schema, which can then be processed by the schemaFieldVisitor.
+	schemaToModel map[string]string
+}
+
+// byIDHandler is the Handler method for a request to get the schema definition for a specifc schema. Will use the
+// cached models found during the last refresh as part of this process.
+func (s *schemaDefinitionHandler) byIDHandler(request *types.APIRequest) (types.APIObject, error) {
+	// pseudo-access check, designed to make sure that users have access to the schema for the definition that they
+	// are accessing.
+	requestSchema := request.Schemas.LookupSchema(request.Name)
+	if requestSchema == nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.NotFound, "no such schema")
+	}
+
+	if s.needsRefresh() {
+		err := s.refresh()
+		if err != nil {
+			logrus.Errorf("error refreshing schemas %s", err.Error())
+			return types.APIObject{}, apierror.NewAPIError(internalServerErrorCode, "error refreshing schemas")
+		}
+	}
+
+	// lock only in read-mode so that we don't read while refresh writes. Only use a read-lock - using a write lock
+	// would make this endpoint only usable by one caller at a time
+	s.RLock()
+	defer s.RUnlock()
+
+	if s.models == nil {
+		return types.APIObject{}, apierror.NewAPIError(notRefreshedErrorCode, "schema definitions not yet refreshed")
+	}
+	models := *s.models
+	modelName, ok := s.schemaToModel[requestSchema.ID]
+	if !ok {
+		return types.APIObject{}, apierror.NewAPIError(notRefreshedErrorCode, "no model found for schema, try again after refresh")
```

Comment:
>  It could also be requestSchema.ID is just completely incorrect, and in that case no amount of waiting will ever help, right?

This part is incorrect. We first [call LookupSchema](https://github.com/rancher/steve/blob/bc3457781eb18d93c4cb2709f2c5940502455178/pkg/schema/definitions/handler.go#L54) to make sure that the schema for this ID exists before we get to this part. In the case that you mentioned this endpoint will return a 404.

This error indicates that while we have a schema for this ID, the openAPI definition that we have cached does not have a record of this value. This can be possible due to the fact that the schemas have a different refresh mechanism than the definitions, which is potentially more up-to-date. The case where this can occur is in the (max 5 seconds) time period between when a new CRD is installed and when the endpoint refreshes the schemas and can pick up the new model.

Reply:
Right, that makes sense, it was my misunderstanding. Resolving.